Added animal overlay to drone filters

coiax opened the pull request:
![A bunch of animals in a room](https://cloud.githubusercontent.com/assets/609465/14664825/3a611c52-06c6-11e6-8dde-f58a85675f49.png)

:cl: coiax
rscadd: Centcom's "drone expert" reports that drones vision filters have
been upgraded to include the option of making all beings appear as harmless animals.
info: As always, all drone filters are togglable.
/:cl:

Inspired by a round where the cleanbot bug made everyone look like a
cat, now the drones ignore everyone, because they're just a bunch of
cats, cows and occasionally butterflies.

Fixes #17057

coiax committed:
Added animal overlay to drone filters

:cl: coiax
rscadd: Centcom's "drone expert" reports that drones vision filters have
been upgraded to include the option of making all beings appear as harmless animals.
info: As always, all drone filters are togglable.
/:cl:

Inspired by a round where the cleanbot bug made everyone look like a
cat, now the drones ignore everyone, because they're just a bunch of
cats, cows and occasionally butterflies.

Core0verload commented:
why

coiax commented:
Because when people tell me that I shouldn't nerf Mulebots, I do drone changes instead.

RandomMarine commented:
Honestly, fuck the vision filter.

There needs to be an off option.

coiax commented:
Vision filter's there to make drones remember to ignore people. Your behaviour does change if it's disabled.

RandomMarine commented:
Drones aren't explicitly meant to outright ignore. The laws dictate not to involve yourself in their matters, but it's important to know at a glance whether or not a group of carp has any that are alive and ready to kill you.

It's important to know at a glance if that human just past a broken window is an actual department worker that won't mind you stopping to fix. Or the chaplain who just broke in who'll whip out his choice of two-hit-kill meme weapon next he sees you.

Having to examine very static blob in sight to make sure it's safe to do your task doesn't add anything to the game. Filter pls go.

TankNut commented:
While I'm not a big fan of the whole vision filter business it's nice to have more modes. Static tends to get annoying after a while and I can't see myself ever using letter, leaving only the blank option.

coiax commented:
Letter is good when you need to tell the difference between humans and zombies.

TheRealGLH commented:
Why is there a pokemon between the animals?

coiax commented:
That would be the original Runtime cat sprite, I believe.

RemieRichards commented on code/__HELPERS/icons.dm:
```diff
@@ -872,6 +872,52 @@ The _flatIcons list is a cache for generated icon files.
+	var/list/icons = list()
+	var/icon/selected
+
+	icons += new/icon('icons/mob/animal.dmi', "cow")
```
can you not just loop through all the subtypes of simple animals and pick one to grab it's appearance, instead of a million /icon objects?

TankNut replied:
You could probably loop through all the simple animal subtypes and check for gold_core_spawnable == 2, which determines 'friendly' animals so that you don't end up with wraiths and faithless and whatever other nasty things everywhere.

coiax replied:
Yeah, but then we wouldn't be able to pick Ian in a wheelchair. If we're concerned about the time creating the icons will take, I can probably ask for advice on making the list global so it only needs to be created once.

RemieRichards replied:
so just add a ```corgi/old``` subtype.

coiax replied:
I don't know if there's any programmatic way of picking just the icons that I want without just listing them. The bolt of change code is similar, it just has a bunch of different mobs that you can become.

coiax replied:
@RemieRichards Remie, why, making mob icons out of friendly gold slime spawns is turning out to be requiring a lot of extra work to make work, especially with chickens because of how they only pick valid icon_state in their New(), and rabbits are also a child class of rabbits, and IT'S SO MUCH WORK. We could make chickens have an initial valid icon_state, but then we lose the ability to have different coloured chickens in filters, same with rabbits, and I've already special cased butterflies, and wew.

RemieRichards replied:
You must not be aware of how shit the ```/icon``` type is, it carries with it quite a high performance cost because it contains all the information of the image state (which is why it's able to be blended/coloured etc), you're making 30 of these, EVERY TIME SOMETHING COMES INTO EXISTENCE.

It may very well be more work, but it makes everything about it better.

coiax replied:
Remie, what if I just made that list once, put it in a global, and then just reused it. Would that be better?

RemieRichards replied:
You're still mulling around with one of the shittest object types DM has.

coiax committed:
Forgot the little lizard

TechnoAlchemisto commented:
doesn't seem necessary

coiax committed:
Humans special case drone overlay for some reason??

coiax committed:
Only make the list once

coiax committed:
Added super call to snowflake drone New()

coiax committed:
Added 'drone_vision' icon to actions.dmi

Used a TV image now, seems like a decent enough icon, unless someone
wants to make something better.

coiax committed:
Action button for verb, random filter choice

coiax committed:
Modified icons, works, more compact

coiax committed:
Merge branch 'master' into drone-sees-you-as-a-big-stupid-cat

Conflicts:
	icons/mob/actions.dmi

coiax committed:
Re-added 'drone_vision' to actions.dmi

coiax committed:
Easy code review changes

coiax committed:
Changes to gold core list of images

This PR is definitely getting squashed amirite

coiax committed:
Merge branch 'drone-sees-you-as-a-big-stupid-cat' of github.com:coiax/-tg-station into drone-sees-you-as-a-big-stupid-cat

coiax committed:
Makes it compile

RemieRichards commented on code/__HELPERS/icons.dm:
```diff
@@ -872,6 +872,28 @@ The _flatIcons list is a cache for generated icon files.
+
+// Pick a random animal instead of the icon, and use that instead
+/proc/getRandomAnimalImage(atom/A)
+	if(friendly_animals.len <= 0)
```
```if(!friendly_animals.len)```

RemieRichards commented on code/__HELPERS/icons.dm:
```diff
@@ -872,6 +872,28 @@ The _flatIcons list is a cache for generated icon files.
+
+
+	var/mob/living/simple_animal/SA = pick(friendly_animals)
+	var/icon/I = icon(initial(SA.icon), initial(SA.icon_state))
```
make this ```/image```, butterflies use ```.color``` and so don't need ```/icon```

coiax committed:
Use of image, not icon

RemieRichards commented on code/__HELPERS/icons.dm:
```diff
@@ -872,6 +872,31 @@ The _flatIcons list is a cache for generated icon files.
+	var/image/final_image = image(icon, icon_state=icon_state, loc = A)
+
+	if(ispath(SA, /mob/living/simple_animal/butterfly))
+		var/color = rgb(rand(0,255), rand(0,255), rand(0,255))
```
this var isn't needed, just do it in one line.

coiax committed:
Color in one line

coiax committed:
Code review fix

coiax committed:
Code review
